Const array_proxy indexing throws linking error

// This works
    const af::array a(100, 100);
    af::array b = a.cols(0, 20);
    b = b.rows(10, 20);

    // This works
    af::array a(100, 100);
    af::array b = a.cols(0, 20).rows(10, 20);

    // This fails
    const af::array a(100, 100);
    af::array b = a.cols(0, 20).rows(10, 20);

The linker error is the following:

    main.cc.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class af::array::array_proxy const __cdecl af::array::array_proxy::rows(int,int)const " (__imp_?rows@array_proxy@array@af@@QEBA?BV123@HH@Z) referenced in function main

Platform: Windows 10, 64-bit, Visual Studio 2013
ArrayFire: 3.0.2

--- a/src/api/cpp/array.cpp
+++ b/src/api/cpp/array.cpp
@@ -739,12 +739,19 @@ namespace af
     }
 
     //FIXME: Check if this leaks
-#define MEM_INDEX(FUNC_SIG, USAGE)          \
-    array::array_proxy                      \
-    array::array_proxy::FUNC_SIG            \
-    {                                       \
-        array *out = new array(this->get());\
-        return out->USAGE;                  \
+#define MEM_INDEX(FUNC_SIG, USAGE)                  \
+    array::array_proxy                              \
+    array::array_proxy::FUNC_SIG                    \
+    {                                               \
+        array *out = new array(this->get());        \
+        return out->USAGE;                          \
+    }                                               \
+                                                    \
+    const array::array_proxy                        \
+    array::array_proxy::FUNC_SIG const              \
+    {                                               \
+        const array *out = new array(this->get());  \
+        return out->USAGE;                          \
     }
 
     MEM_INDEX(row(int index)                , row(index));

--- a/src/api/cpp/array.cpp
+++ b/src/api/cpp/array.cpp
@@ -606,6 +606,14 @@ namespace af
         return out;
     }
 
+    af_array array::array_proxy::get() const
+    {
+        array tmp = *this;
+        af_array out = 0;
+        AF_THROW(af_retain_array(&out, tmp.get()));
+        return out;
+    }
+
 #define MEM_FUNC(PREFIX, FUNC)                  \
     PREFIX array::array_proxy::FUNC() const     \
     {                                           \
@@ -613,7 +621,6 @@ namespace af
         return out.FUNC();                      \
     }
 
-    MEM_FUNC(af_array               , get)
     MEM_FUNC(dim_t                  , elements)
     MEM_FUNC(array                  , T)
     MEM_FUNC(array                  , H)

--- a/test/array.cpp
+++ b/test/array.cpp
@@ -373,3 +373,19 @@ TEST(Array, ShapeAttributes)
     EXPECT_FALSE(volume.    iscolumn());
     EXPECT_FALSE(hypercube. iscolumn());
 }
+
+TEST(Array, ISSUE_951)
+{
+// This works
+    //const af::array a(100, 100);
+    //af::array b = a.cols(0, 20);
+    //b = b.rows(10, 20);
+
+// This works
+    //af::array a(100, 100);
+    //af::array b = a.cols(0, 20).rows(10, 20);
+
+// This fails with linking error
+    const af::array a = randu(100, 100);
+    af::array b = a.cols(0, 20).rows(10, 20);
+}

